core/src/main/java/io/grpc/internal/MessageFramer.java

```diff
@@ -163,18 +170,27 @@ public void writePayload(InputStream message) {
 
   private int writeUncompressed(InputStream message, int messageLength) throws IOException {
     if (messageLength != -1) {
-      currentMessageWireSize = messageLength;
       return writeKnownLengthUncompressed(message, messageLength);
     }
     BufferChainOutputStream bufferChain = new BufferChainOutputStream();
-    int written = writeToOutputStream(message, bufferChain);
+    int written;
+    if (streamCompression) {
+      // "Full stream" compression with unknown length reduces to separate concatenated gzip blocks
```

Comment:
We want to reuse the gzip context, so if we need to "eat" the copy, so be it. It may be possible to avoid the copy by using the deflater directly, but let's not worry about it at this point. We will probably want to use a byte[]-backed buffer chain so that we can read from it, however.

Note that this code doesn't look "safe" as-is because `gzipOutputStream` hasn't been finished and retains its window, which will be wrong for the receiver.